File: structural/dart/proxy_structured/logging_proxy.dart
abstract class Service {
  void doWork();
}

class RealService extends Service {
  @override
  void doWork() {
    print('Service: Performing work...');
  }
}

class LoggingProxy extends Service {
  final RealService _realService = RealService();
  
  void _log(String message) {
    print('[LOG] $message');
  }
  
  @override
  void doWork() {
    _log('Before doWork');
    _realService.doWork();
    _log('After doWork');
  }
}

File: structural/typescript/proxy_structured/logging_proxy.ts
export interface Service {
  doWork(): void;
}

export class RealService implements Service {
  doWork(): void {
    console.log('Service: Performing work...');
  }
}

export class LoggingProxy implements Service {
  private realService: RealService;
  
  constructor() {
    this.realService = new RealService();
  }
  
  private log(message: string): void {
    console.log(`[LOG] ${message}`);
  }
  
  doWork(): void {
    this.log('Before doWork');
    this.realService.doWork();
    this.log('After doWork');
  }
}
